clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -3062,6 +3062,17 @@ void SemaHLSL::ActOnVariableDeclarator(VarDecl *VD) {
     if (VD->getType()->isHLSLIntangibleType())
       collectResourceBindingsOnVarDecl(VD);
 
+    const Type *VarType = VD->getType().getTypePtr();
+    if (VarType->isArrayType()) {
+      VarType = VarType->getArrayElementTypeNoTypeQual();
+    }
```

Comment:
```suggestion
    while (VarType->isArrayType())
      VarType = VarType->getArrayElementTypeNoTypeQual();
```
There might be multi-dimensional arrays of resources.